lingpy/sequence/ngrams.py

```diff
@@ -0,0 +1,1441 @@
+            lentropy += [-p*math.log(p, 2.0) for p in _probs]
+
+        return sum(lentropy)
+
+    def entropy(self, sequence, base=2.0):
+        """
+        Calculates the cross-entropy of a sequence.
+
+        Parameters
+        ----------
+        sequence: list
+            The sequence whose cross-entropy will be calculated.
+
+        base: float
+            The logarithmic base for the cross-entropy calculation. Defaults to
+            2.0, following the standard approach set by Shannon that allows
+            to consider entropy in terms of bits needed for unique
+            representation.
+
+        Returns
+        -------
+        ch: float
+            The cross-entropy calculated for the sequence, a real number.
+        """
+        return -(self.score(sequence)/math.log(base)) / len(sequence)
+
+    def perplexity(self, sequence):
+        """
+        Calculates the perplexity of a model.
+
+        As per definition, this is simply 2.0 to the cross-entropy of the
+        given sequence on logarithmic base of 2.0.
+
+        Parameters
+        ----------
+        sequence: list
+            The sequence whose perplexity should be calculated.
+
+        Returns
+        -------
+        perplexity: float
+            The calculated perplexity for the sequence.
+        """
+        return 2.0 ** self.entropy(sequence)
+
+
+    def save_model(self, filename):
+        """
+        Saves an ngram model to disk.
+
+        This method serializes a model, either trained or not, to a binary
+        format, allowing future reuse without the time and computation
+        costs. The model is supposed to be loaded with the `.load_model()`
+        method.
+
+        Parameters
+        ----------
+        filename: str
+            The path to the file where to write the model.
+        """
+
+        # We must use pickle and not a more readable format such as JSON as
+        # the latter does not allow for the serialization of tuples (unless
+        # everything is mapped back and forth to and from strings). Even
+        # pickle does not allow for all data structures we use (in particular,
+        # those with lambdas), so that we need to cast back and forth to
+        # primitive datatypes when saving and loading.
+        model_data = {
+            '_padsymbol' : self._padsymbol,
+            '_pre' : self._pre,
+            '_post' : self._post,
+            '_ngrams' : dict(self._ngrams),
+            '_ngram_space' : self._ngram_space,
+            '_seqlens' : self._seqlens,
+            '_smooth_method' : self._smooth_method,
+            '_bins' : self._bins,
+            '_smooth_kwargs' : self._smooth_kwargs,
+            '_trained' : self._trained,
+            '_p' : self._p,
+            '_p0' : self._p0,
+            '_l' : self._l,
+            '_l0' : self._l0,
+        }
+
+        with open(filename, 'wb') as handler:
+            pickle.dump(model_data, handler, pickle.HIGHEST_PROTOCOL)
```

Comment:
I'm a bit on the fence regarding this pickling. It's really hard to support object persistence in this way transparently for the end user. What happens, e.g. when additional non-picklable attributes are added? Then old pickles won't survive. So should a version be added to the pickles? Should there be support to migrate old pickles? Also pickles are not portable between py2 and py3.

I'm leaning towards delegating things like object persistence to the end user - maybe adding a method (or property) here which returns a picklable dict representation of the object (and allowing `__init__` to work with such a representation). The actual pickling/unpickling, i.e. interfering with the files, should be left to the end user. Maybe they have a [ZODB](https://en.wikipedia.org/wiki/Zope_Object_Database) running somewhere and want to use that for persitence ...

Reply:
Yes, I agree. We had similar problems with support vector machines: they take ages to be trained, but then, they could only be stored in pickle, which renders them useless, so all users have to run their computer for two weeks in order to be able to use the cognate detection method, although all they create in the end is a matrix.

Reply:
I see your point and agree with it. I also hated to use `pickle`, and at first wanted to use at least `json` or `shelve`.

I implemented these methods because they were indirectly requested by the reviewers of a paper, as they were requesting many models trained on many subsets of words (each interaction was taking almost 30 minutes in my laptop). "My" problem is that the end-user here likely wants simple and stupid "save" and "load" methods, and any additional code could be taken as an unacceptable hack (even if the library does *exactly* the same under the hood).

That said, using ZODB is an overkill and no other solution is really good. I will remove these methods and (re)implement them in a class that extends NgramModel (maybe even putting in in third-party later) which we can use in case of submitting code for review. Leaner code in lingpy.